### PR TITLE
[WIP] owdbscan: the first version of the DBSCAN widget

### PR DESCRIPTION
**Pull request by BlazZupan**
The first version of the new clustering widget that implements DBSCAN algorithm. The widget has data on the input and on the output augments this data set with column for a cluster and a column indicating if the data instance is an essential member of the cluster. Only three distance functions are included, as others, for some unknown reason, proved to be very slow (something to be yet investigated why). The output should be nicely rendered in scatterplot, once this is corrected such that it also shows data with unknown class and the size of the point could be related to a meta variable (a task I asked for Aleš's help yesterday).

### CODE REVIEW

File: Orange/widgets/unsupervised/owdbscan.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    def check_data_size(self):
+        if len(self.data) < 2:
+            self.error("Not enough unique data instances" +
+                       "({}).".format(len(self.data)))
+            return False
+        return True
+
+    def commit(self):
+        print("RUN")
```
Comment (kernc) [inline Orange/widgets/unsupervised/owdbscan.py]: This line probably needs to go.

File: Orange/widgets/unsupervised/owdbscan.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        self.layout().setSizeConstraint(QtGui.QLayout.SetFixedSize)
+
+    def adjustSize(self):
+        self.ensurePolished()
+        self.resize(self.leftWidgetPart.sizeHint())
+
+    def check_data_size(self):
+        if len(self.data) < 2:
+            self.error("Not enough unique data instances" +
+                       "({}).".format(len(self.data)))
```
Comment (kernc) [inline Orange/widgets/unsupervised/owdbscan.py]: You needn't a `+` to concatenate strings like that. It works as in C where neighbouring strings are combined.
```py
self.error("Not enough unique data instances "
           "({})".format(len(self.data)))
```

File: Orange/widgets/unsupervised/owdbscan.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        if not self.check_data_size():
+            return
+        self.model = DBSCAN(
+            eps=self.eps,
+            min_samples=self.min_samples
+        )(self.data)
+        self.send_data()
+
+    def send_data(self, row=None):
+        print("SEND")
```
Comment (kernc) [inline Orange/widgets/unsupervised/owdbscan.py]: And this.

File: Orange/widgets/unsupervised/owdbscan.py (hunk 1 above)
Comment (lanzagar) [inline Orange/widgets/unsupervised/owdbscan.py]: Remove debug line.

File: Orange/widgets/unsupervised/owdbscan.py (hunk 3 above)
Comment (lanzagar) [inline Orange/widgets/unsupervised/owdbscan.py]: Remove debug line.

File: Orange/widgets/unsupervised/owdbscan.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        box = gui.widgetBox(self.controlArea, "Parameters")
+        gui.spin(box, self, "min_samples", 1, 100, 1, callback=self._invalidate,
+                 label="Core point neighbors")
+        gui.doubleSpin(box, self, "eps", 0.1, 0.9, 0.1,
+                       callback=self._invalidate,
+                       label="Neighborhood distance")
+
+        box = gui.widgetBox(self.controlArea, self.tr("Distance Metric"))
+        gui.comboBox(box, self, "metric_idx",
+                     items=list(zip(*_METRICS))[0],
```
Comment (lanzagar) [inline Orange/widgets/unsupervised/owdbscan.py]: The selected metric is never used

File: Orange/widgets/unsupervised/owdbscan.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+        self.send("Annotated Data", new_table)
+
+    def set_data(self, data):
+        self.data = data
+        if self.data is None:
+            self.send("Annotated Data", None)
+        self.commit()
+
+    def _invalidate(self):
```
Comment (lanzagar) [inline Orange/widgets/unsupervised/owdbscan.py]: Why is this method needed?
Calling commit directly instead of calling _invalidate would be the same...

Comment (astaric) [inline Orange/widgets/unsupervised/owdbscan.py]: auto_commit button replaces commit method with its own (one that checks the autocommit flag). If components reference self.commit as a callback, they would perform unconditional commits.

**Pull request closed without merging**